<?xml version='1.0'?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="robot_car_seb">

    <!-- PROPIEDADES -->
    <xacro:property name="blx" value="${10/10}" />
    <xacro:property name="bly" value="${8/10}" />
    <xacro:property name="blz" value="${5/10}" />

    <xacro:property name="m" value="1" />

    <xacro:property name="olx" value="${2.5/10}" />
    <xacro:property name="oly" value="${(4 + 1.25/2)/10}" />
    <xacro:property name="olz" value="${1.25/10}" />

    <xacro:property name="crad" value="${2.5/10}" />
    <xacro:property name="lrad" value="${1.25/10}" />

    <xacro:property name="srad" value="${(2.5/2)/10}" />
    <xacro:property name="slx" value="${10/10 - srad}" />


    <!-- MACROS / FUNCIONES -->
    <xacro:macro name="wheel_macro" params="p_name p_radius p_length">
        <link name="${p_name}">
            <visual>
                <!-- box geometry -->
                <geometry>
                    <cylinder radius="${p_radius}" length="${p_length}" />
                </geometry>
                <origin xyz="0 0 0" rpy="${-pi/2} 0 0" />
            </visual>

            <inertial>
                <mass value="${m}" />
                <origin xyz="0 0 0" rpy="0 0 0" />
                <inertia ixx="${1/12*m*(p_length*p_length + 3*p_radius*p_radius)}" ixy="0" ixz="0"
                    iyy="${1/12*m*(p_length*p_length + 3*p_radius*p_radius)}" iyz="0"
                    izz="${1/2*m*p_radius*p_radius}" />
            </inertial>

            <collision>
                <!-- box geometry -->
                <geometry>
                    <cylinder radius="${p_radius}" length="${p_length}" />
                </geometry>
                <origin xyz="0 0 0" rpy="${-pi/2} 0 0" />
            </collision>
        </link>
    </xacro:macro>


    <!-- INICIO ROBOT -->
    <link name="base_link">
        <!-- ... -->
    </link>

    <joint name="base_chassis_joint" type="fixed">
        <parent link="base_link" />
        <child link="chassis_link" />
        <!--Crea offsets the child con respecto a parent -->
        <origin xyz="${-olx} 0 ${-olz}" rpy="0 0 0" />
    </joint>

    <link name="chassis_link">
        <visual>
            <!-- box geometry -->
            <geometry>
                <box size="${blx} ${bly} ${blz}" />
            </geometry>
            <!-- Esto mueve la caja desde el centro, no el origen con respecto a la caja-->
            <origin xyz="${blx/2} 0 ${blz/2}" rpy="0 0 0" />
        </visual>

        <inertial>
            <mass value="${m}" />
            <origin xyz="0 0 0" rpy="0 0 0" />
            <inertia ixx="${1/12*m*(blz*blz + bly*bly)}" ixy="0" ixz="0"
                iyy="${1/12*m*(blx*blx + blz*blz)}" iyz="0" izz="${1/12*m*(bly*bly + blx*blx)}" />
        </inertial>

        <collision>
            <!-- box geometry -->
            <geometry>
                <box size="${blx} ${bly} ${blz}" />
            </geometry>
            <origin xyz="${blx/2} 0 ${-blz/2}" rpy="0 0 0" />
        </collision>
    </link>
    <!-- ... -->
    <xacro:wheel_macro p_name="rwheel_link"
        p_radius="${crad}"
        p_length="${lrad}" />

    <joint name="chassis_rwheel_joint" type="continuous">
        <parent link="chassis_link" />
        <child link="rwheel_link" />
        <origin xyz="${olx} ${oly} ${olz}" rpy="0 0 0" />
        <axis xyz="0 1 0" />
        <limit effort="1" lower="0" upper="${2*pi}" velocity="1" />
        <dynamics damping="1" friction="1" />
    </joint>

    <xacro:wheel_macro p_name="lwheel_link"
        p_radius="${crad}"
        p_length="${lrad}" />


    <joint name="chassis_lwheel_joint" type="continuous">
        <parent link="chassis_link" />
        <child link="lwheel_link" />
        <origin xyz="${olx} ${-oly} ${olz}" rpy="0 0 0" />
        <axis xyz="0 1 0" />
        <limit effort="1" lower="0" upper="${2*pi}" velocity="1" />
        <dynamics damping="1" friction="1" />
    </joint>


    <link name="sphere_link">
        <visual>
            <!-- Sphere-->
            <geometry>
                <sphere radius="${srad}" />
            </geometry>
            <origin xyz="0 0 0" rpy="0 0 0" />
        </visual>

        <inertial>
            <mass value="${m}" />
            <origin xyz="0 0 0" rpy="0 0 0" />
            <inertia ixx="${m*2/5*srad*srad}" ixy="0" ixz="0"
                iyy="${m*2/5*srad*srad}" iyz="0" izz="${m*2/5*srad*srad}" />
        </inertial>

        <collision>
            <!-- Sphere-->
            <geometry>
                <sphere radius="${srad}" />
            </geometry>
            <origin xyz="0 0 0" rpy="0 0 0" />
        </collision>
    </link>

    <joint name="chassis_sphere_joint" type="fixed">
        <parent link="chassis_link" />
        <child link="sphere_link" />
        <origin xyz="${slx} 0 0" rpy="0 0 0" />
    </joint>


</robot>